crates/ruff_python_semantic/src/analyze/class.rs

```diff
@@ -2,29 +2,26 @@ use rustc_hash::FxHashSet;
 
 use ruff_python_ast as ast;
 use ruff_python_ast::helpers::map_subscript;
-use ruff_python_ast::name::QualifiedName;
+use ruff_python_ast::Expr;
 
 use crate::{BindingId, SemanticModel};
 
 /// Return `true` if any base class matches a [`QualifiedName`] predicate.
 pub fn any_qualified_name(
     class_def: &ast::StmtClassDef,
     semantic: &SemanticModel,
-    func: &dyn Fn(QualifiedName) -> bool,
+    func: &dyn Fn(&Expr) -> bool,
```

Comment:
I had to change this to make it more flexible, which in turn required changing all the call sites mechanically.

Reply:
It may be worth to introduce a new method instead:

```rust
pub fn any_base_class(class_def: &ast::StmtClassDef, semantic: &SemanticModel, func: &dyn Fn(&Expr) -> bool) -> bool {
	// The body of `any_qualified_name` after your refactor
	...
}


pub fn any_qualified_base_class(class_def, &ast::StmtClassDef, semantic: &SemanticModel, func: &dyn Fn(QualifiedName) -> bool) -> bool {
	any_base_class(class_def, semantic, |expr| semantic.resolve_qualified_name(map_subscript(expr)).is_some_and_func(func))
}
```

Reply:
+1 for Micha's suggestion. This change otherwise makes most callsites a lot more verbose.

One _small advantage_ of the way Charlie currently has it is that it means that the caller is forced to consider whether the bases could actually be subscript expressions or not. In a lot of situations it's not strictly correct to use `map_subscript` on a class base, but the previous function was doing it unconditionally on all class bases it iterated over. But I just don't think the theoretical correctness issue there is that significant, really.

Reply:
Honestly I don't agree -- it's actually way more consistent (and flexible) for us to pass back the expression, and we use `resolve_qualified_name` for all related checks elsewhere -- but I also don't feel strongly. I'll just change it.

Reply:
I don't feel too strongly about this point either tbh. It just seems a shame that so many callsites get more verbose when we only need the extra flexibility in one instance (currently)